### ISSUE TITLE
Issue with Alamore challenge delegate and escaping closure

### ISSUE THREAD

**Issue by nitinsh99** [Alamofire/Alamofire#2108]
I am using AF and using it's delegate to catch the authentication chanllenge returned by my server. 

    func connectGetRequest(_ url : URL){

        let sessionManager = Alamofire.SessionManager.default
        sessionManager.request(url).responseString { response in
            print("Response String: \(response.result.value)")
        }
        let delegate: Alamofire.SessionDelegate = sessionManager.delegate
        
        
        delegate.taskDidReceiveChallengeWithCompletion = { session, task, challenge,  completionHander in
            print("session is \(session), task is \(task) challenge is \(challenge.protectionSpace.authenticationMethod) and handler is \(completionHander)")
            if(challenge.protectionSpace.authenticationMethod == "NSURLAuthenticationMethodServerTrust"){

                completionHander(.performDefaultHandling,nil)
            }else{
                
                print("chanllege type is \(challenge.protectionSpace.authenticationMethod)")
                
          // Following line give me the error: "passing non-escaping parameter 'completionHander' to function expecting an @escaping closure"

    self.handleAuthenticationforSession(challenge,completionHandler:  completionHander)
            }
        }
        
        delegate.dataTaskDidReceiveData = {session , task, data in
            
            print("received data \(data)")
        }
        
    }

    
    func handleAuthenticationforSession(_ challenge: URLAuthenticationChallenge,completionHandler:   @escaping (Foundation.URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        
             
            Authhandler.handleChallenge(forURLSessionChallenge: challenge, completionHandler: completionHandler)
            

    }
    

issue I have:

 
 1. If I use the code above as it is, I get

  error: "passing non-escaping  parameter 'completionHander' to function expecting an @escaping closure"

 2. If I make the parameter of the function handleAuthenticationSession non escaping, I get :

         func handleAuthenticationforSession(_ challenge: URLAuthenticationChallenge,
          completionHandler:   (Foundation.URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        
             }

error: "Closure use of non-escaping parameter 'completion' may allow it to escape"

So basically I am stuck in a deadlock while I use Alamofire's closure syntax to delegate the auth challenge.

**Comment by nitinsh99** [comment]
Update:

Someone on SO pointed out that I am hitting the same issue as [this](https://github.com/Alamofire/Alamofire/pull/1951) bug.

Can  someone please confirm?
Thanks

**Comment by jshier** [comment]
Yes, that's a bug that will be fixed as part of our 4.5 release.
